eclair-core/src/main/scala/fr/acinq/eclair/io/IncomingConnectionsTracker.scala

```diff
@@ -0,0 +1,80 @@
+import fr.acinq.eclair.io.Peer.Disconnect
+
+/**
+ * A singleton actor that limits the total number of incoming connections from peers that do not have channels with us.
+ *
+ * When a new incoming connection request is received, the Switchboard should send an
+ * [[IncomingConnectionsTracker.TrackIncomingConnection]] message.
+ *
+ * When the number of tracked peers exceeds `eclair.peer-connection.max-no-channels`, send [[Peer.Disconnect]] to
+ * the tracked peer with the oldest incoming connection.
+ *
+ * When a tracked peer disconnects or confirms a channel, we will stop tracking that peer.
+ *
+ * We do not need to track peers that disconnect because they will terminate if they have no channels.
+ * Likewise, peers with channels will disconnect and terminate when their last channel closes.
+ *
+ * Note: Peers on the sync whitelist are not tracked.
+
+ * This actor enables a DoS attack that can block new incoming liquidity from non-whitelisted nodes.
+ * An attacker can trivially disconnect other incoming connections before they can open channels by repeatedly opening
+ * new connections using random node IDs.
+*/
+object IncomingConnectionsTracker {
+  // @formatter:off
+  sealed trait Command
+
+  case class TrackIncomingConnection(remoteNodeId: PublicKey) extends Command
+  private case class ForgetIncomingConnection(remoteNodeId: PublicKey) extends Command
+  private[io] case class CountIncomingConnections(replyTo: ActorRef[Int]) extends Command
+  // @formatter:on
+
+  def apply(nodeParams: NodeParams, switchboard: ActorRef[Disconnect]): Behavior[Command] = {
+    Behaviors.setup { context =>
+      context.system.eventStream ! EventStream.Subscribe(context.messageAdapter[PeerDisconnected](c => ForgetIncomingConnection(c.nodeId)))
+      context.system.eventStream ! EventStream.Subscribe(context.messageAdapter[ChannelOpened](c => ForgetIncomingConnection(c.remoteNodeId)))
+      new IncomingConnectionsTracker(nodeParams, switchboard, context).tracking(Map())
+    }
+  }
+}
+
+private class IncomingConnectionsTracker(nodeParams: NodeParams, switchboard: ActorRef[Disconnect], context: ActorContext[Command]) {
+  import IncomingConnectionsTracker._
+
+  private def tracking(incomingConnections: Map[PublicKey, TimestampMillis]): Behavior[Command] = {
+    Metrics.IncomingConnectionsNoChannels.withoutTags().update(incomingConnections.size)
+    Behaviors.receiveMessage {
+      case TrackIncomingConnection(remoteNodeId) =>
+        if (nodeParams.syncWhitelist.contains(remoteNodeId)) {
+          Behaviors.same
+        } else {
+          if (incomingConnections.size >= nodeParams.peerConnectionConf.maxNoChannels) {
+            Metrics.IncomingConnectionsDisconnected.withoutTags().increment()
+            val oldest = incomingConnections.minBy(_._2)._1
+            context.log.warn(s"disconnecting peer=$oldest, too many incoming connections from peers without channels.")
+            switchboard ! Disconnect(oldest)
```

Comment:
This creates an issue if the `Switchboard` cannot find that `Peer`: the switchboard will send back a `Status.Failure` to the `IncomingConnectionsTracker`, which will crash the actor because it cannot convert it to a `IncomingConnectionsTracker.Command` :cry: 

Maybe we should update `Disconnect` to take a `boolean` argument to indicate whether we expect a response or not? It looks like most of the codebase actually just wants a fire-and-forget message without a response, the only exception is when we call this from the API.

Reply:
Fixed in [be5366a](https://github.com/ACINQ/eclair/pull/2601/commits/be5366a4a35f4c13f0f6499f8ad41e78e37d2fa8). I took the second cleaner fix option, but for consistency, `None` replies to the `sender()` as before and I pass `Some(context.system.ignoreRef[DisconnectResponse])` as the `replyTo_opt` parameter in `IncomingConnectionTracker` to prevent receiving an unhandled reply.